var/spack/repos/builtin/packages/qt-base/package.py

```diff
@@ -0,0 +1,115 @@
+    variant('ssl', default=True, description='Build with OpenSSL support.')
+    variant('widgets', default=True, description='Build with widgets.')
+
+    depends_on('cmake@3.16:', type='build')
+    depends_on('ninja', type='build')
+    depends_on("pkgconfig", type='build')
+    depends_on("python", type='build')
+
+    # Dependencies, then variant- and version-specific dependencies
+    depends_on('at-spi2-core', when='+accessibility')
+    depends_on('dbus', when='+dbus')
+    depends_on('double-conversion')
+    depends_on('fontconfig')
+    depends_on("freetype")
+    depends_on('gl', when='+opengl')
+    depends_on("harfbuzz")
+    depends_on("icu4c")
+    depends_on("jpeg")
+    depends_on('libdrm')
+    depends_on('libjpeg')
+    depends_on("libmng")
+    depends_on('libproxy')
+    depends_on("libtiff")
+    depends_on('libxkbcommon')
+    depends_on("libxml2")
+    depends_on('libxrender')
+    depends_on('openssl', when='+ssl')
+    depends_on('pcre2+multibyte')
+    depends_on("sqlite", when='+sql')
+    depends_on("zlib")
+    depends_on("zstd")
+
+    def patch(self):
+        import os
+        import shutil
+        vendor_dir = join_path(self.stage.source_path, 'src/3rdparty')
+        vendor_deps_to_keep = [
+            'blake2', 'easing', 'forkfd', 'freebsd',
+            'icc', 'md4', 'md4c', 'md5', 'rfc6234',
+            'sha1', 'sha3', 'tinycbor', 'VulkanMemoryAllocator',
+        ]
+        with working_dir(vendor_dir):
+            for dep in os.listdir():
+                if os.path.isdir(dep):
+                    if dep not in vendor_deps_to_keep:
+                        shutil.rmtree(dep)
+
+    def cmake_args(self):
+        def define_feature(variant):
+            return self.define_from_variant('FEATURE_' + variant, variant)
+        args = [
+            self.define_from_variant('BUILD_SHARED_LIBS', 'shared'),
+            self.define_from_variant('QT_BUILD_EXAMPLES', 'examples'),
+            self.define('QT_BUILD_TESTS', self.run_tests),
+            self.define('FEATURE_optimize_size',
+                        self.spec.satisfies('build_type=MinSizeRel')),
+            define_feature('accessibility'),
+            define_feature('dbus'),
+            define_feature('framework'),
+            define_feature('gui'),
+            define_feature('ssl'),
+            define_feature('widgets'),
+            define_feature('sql'),
+        ]
+
+        # INPUT_* arguments: link where possible
+        for x in ['dbus', 'openssl']:
+            args.append(self.define('INPUT_' + x, 'linked'))
+        # But use opengl
+        args.append(self.define_from_variant('INPUT_opengl', 'opengl'))
+
+        # FEATURE_system_* arguments: use system where possible
+        for x in ['doubleconversion', 'freetype', 'harfbuzz', 'jpeg',
+                  'pcre2', 'png', 'proxies', 'sqlite', 'zlib']:
+            args.append(self.define('FEATURE_system_' + x, True))
```

Comment:
Installing with everything off:
```
CMake Warning:
  Manually-specified variables were not used by the project:

    FEATURE_accessibility
    FEATURE_system_freetype
    FEATURE_system_harfbuzz
    FEATURE_system_jpeg
    FEATURE_system_png
    FEATURE_system_sqlite
    FEATURE_system_textmarkdownreader
    INPUT_opengl
```

suggesting those dependencies aren't needed.

Reply:
Maybe something like this? (untested)
```python
        # FEATURE_system_* arguments: use system where possible
        def set_feature_system_flags(feature_system, value):    
            for flag, variant in feature_system:
                if variant is None or self.satisfies(variant):
                    args.append(self.define('FEATURE_system_' + flag, value))   
        feature_system_true = {
            'doubleconversion': None,
            'freetype': '+gui',
            'harfbuzz': '+gui',
            'jpeg': '+gui',
            'pcre2': None,
            'png': '+gui',
            'proxies': None,
            'sqlite': '+sql',
            'zlib': None,
        }
        feature_system_false = {
            'libb2': None,
            'textmarkdownreader': None,
        }
        set_feature_system_flags(feature_system_true, True)
        set_feature_system_flags(feature_system_false, False)
```
Could possibly be even more generic to reuse for other lists that are variant dependent.

Reply:
Maybe—the key thing is that its output gives us a list of `depends_on` that should be conditional based on gui/widgets.